src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -466,14 +467,24 @@ private unsafe void InnerStartOperationReceiveMessageFrom()
                 }
                 _wsaRecvMsgWSABufferArray[0].Pointer = _ptrSingleBuffer;
                 _wsaRecvMsgWSABufferArray[0].Length = _count;
-                _wsaRecvMsgWSABufferArrayGCHandle = GCHandle.Alloc(_wsaRecvMsgWSABufferArray, GCHandleType.Pinned);
-                _ptrWSARecvMsgWSABufferArray = Marshal.UnsafeAddrOfPinnedArrayElement(_wsaRecvMsgWSABufferArray, 0);
+                wsaRecvMsgWSABufferArray = _wsaRecvMsgWSABufferArray;
             }
             else
             {
-                // Just pin the multi-buffer WSABuffer.
-                _wsaRecvMsgWSABufferArrayGCHandle = GCHandle.Alloc(_wsaBufferArray, GCHandleType.Pinned);
-                _ptrWSARecvMsgWSABufferArray = Marshal.UnsafeAddrOfPinnedArrayElement(_wsaBufferArray, 0);
+                // Use the multi-buffer WSABuffer.
+                wsaRecvMsgWSABufferArray = _wsaBufferArray;
+            }
+
+            // Ensure the array is pinned.
+            if (!_wsaRecvMsgWSABufferArrayGCHandle.IsAllocated)
+            {
+                _wsaRecvMsgWSABufferArrayGCHandle = GCHandle.Alloc(wsaRecvMsgWSABufferArray, GCHandleType.Pinned);
+                _ptrWSARecvMsgWSABufferArray = Marshal.UnsafeAddrOfPinnedArrayElement(wsaRecvMsgWSABufferArray, 0);
+            }
+            else
+            {
+                Debug.Assert(_wsaRecvMsgWSABufferArrayGCHandle.Target == wsaRecvMsgWSABufferArray);
```

Comment:
If they didn't that'd be a big problem.  The code is expecting to be able to use _ptrWSARecvMsgWSABufferArray, which is the address of wsaRecvMsgWSABufferArray, which is pinned by _wsaRecvMsgWSABufferArrayGCHandle.  If the handle is allocated but points to something else, then we're no longer pinning what we expected to be pinning.  If the buffer changes, the GCHandle should be freed at that point such that we'll hit the if block rather than the else.